### ISSUE TITLE
Half-edges are not always paired with an opposite.

### ISSUE THREAD

**Issue by olson-sean-k** [olson-sean-k/plexus#8]
Half-edges are only paired with an opposite when a face is present on each side. Half-edges have no opposite along boundaries (where a boundary is an edge belonging to a face with no opposing face). In open meshes, this breaks circulators, which rely on traversing opposing edges. What's worse, there is no indication that something has gone wrong. Given a `Mesh` called `mesh` that forms a single triangle, only the interior of the face would have half-edges. This code would fail:

```
// Get a vertex in the triangle.
let vertex = mesh.vertices().nth(0).unwrap();
// Today, each vertex would have exactly one incoming edge, but this would fail
// with a count of zero.
assert_eq!(1, vertex.incoming_edge().count());
```

Circulators should always work as expected or at least indicate failure. The correct thing to do is probably to represent **every** edge with opposing half-edges, not only when opposing faces are present. It may also be helpful to track boundary edges.

**Comment by olson-sean-k** [comment]
This is a work in progress and there are some important and substantial changes. It will be removed in the future, but see the [edge branch](https://github.com/olson-sean-k/plexus/compare/edge) for details.

Invariants have changed. Not only must half-edges always have opposites, but the connectivity of edges must always be consistent. This means that topological views like `VertexRef` and `EdgeMut` must present a different API to users of the crate, because functions like `into_opposite_edge` should not return an `Option`. Instead, the underlying `Option` should be unwrapped, as a panic indicates a bug. However, views are useful for internal manipulation as well; should there be some sort of internal view API?

Because some connectivity invariants cannot be maintained during bulk operations, a "batch mutation" API has been introduced which defers certain types of errors until the mutation is committed. This allows `from_raw_buffers` and `from_indexer` to function when, for example, the ordering of face insertions causes temporary non-manifold structures to be inserted into the mesh. This has all been moved into a `mutation` module, along with typical immediate mutations. The value of the immediate mutation API is questionable though and it is a bit cumbersome.

Plenty of questions need to be answered here. The current branch is a decent working example, but some APIs will need to be refined. In general, a lot of code in the `graph` module reaches into the guts of the `Mesh` type; that should probably be changed or reduced.

**Comment by olson-sean-k** [comment]
This is fixed by e548625. That change kicked up a lot of dust, and both internal and external APIs are a bit messier than before. The biggest changes include:

- `Mesh` mutations are performed using the `mutation` module. Importantly, it provides a `BatchMutation` that defers certain types of errors until a mutation is committed. This allows from_raw_buffers and from_indexer to function when, for example, the ordering of face insertions causes temporary non-manifold structures to be inserted into the mesh.
- `Mesh`es always have composite edges (every half-edge always has an opposite half-edge). Additionally, all edges should be fully connected. Importantly, this means that circulators are reliable.
- Fewer typologies are supported, though Plexus never promised to support anything non-manifold anyway.

There will likely be some follow-up issues, but e548625 is enough to close this, I think.
